openmoc/materialize.py

```diff
@@ -355,3 +360,328 @@ def load_openmc_mgxs_lib(mgxs_lib, geometry=None):
+
+    # Collect those domains for which we will apply SPH factors
+    if domains == 'all':
+        sph_domains = all_domains
+        sph_indices = np.arange(len(sph_domains))
+    elif domains == 'fissionable':
+        sph_domains = {}
+        sph_indices = []
+        for i, openmc_domain in enumerate(mgxs_lib.domains):
+            if all_domains[openmc_domain.id].isFissionable():
+                sph_domains[openmc_domain.id] = all_domains[openmc_domain.id]
+                sph_indices.append(i)
+    else:
+        py_printf('ERROR', 'SPH factors cannot be applied for domain type '
+                           '%s which is not supported', str(domains))
+
+    # Initialize array of domain-averaged fluxes and SPH factors
+    openmoc_fluxes = np.zeros((num_domains, num_groups))
+
+    # Sore starting verbosity log level
+    log_level = openmoc.get_log_level()
+
+    # Outer SPH loop over domains of interest
+    for i in range(max_domain_iters):
+        for domain_id in sph_domains:
+            py_printf('NORMAL', 'SPH outer iteration %d: %s %d',
+                                i, mgxs_lib.domain_type.capitalize(), domain_id)
+
+            inner_sph = np.ones((num_domains, num_groups))
+
+            for j in range(max_fix_src_iters):
+
+                # Run fixed source calculation with suppressed output
+                if throttle_output and log_level <= openmoc.WARNING:
+                    openmoc.set_log_level('WARNING')
+
+                # Fixed source calculation
+                solver.computeFlux()
+
+                # Restore log output level
+                if throttle_output and log_level <= openmoc.NORMAL: 
+                    openmoc.set_log_level('NORMAL')
+
+                # Extract the FSR scalar fluxes
+                fsr_fluxes = get_scalar_fluxes(solver)
+
+                # Compute the domain-averaged flux in each energy group
+                for k, domain in enumerate(mgxs_lib.domains):
+                    domain_fluxes = fsr_fluxes[fsrs_to_domains == domain.id, :]
+                    openmoc_fluxes[k, :] = np.mean(domain_fluxes, axis=0)
+
+                    # Store index for the domain in the array of SPH factors
+                    if domain.id == domain_id:
+                        sph_index = k
+
+                # Compute SPH factors
+                old_inner_sph = np.copy(inner_sph)
+                inner_sph = openmc_fluxes / openmoc_fluxes
+                inner_sph = np.nan_to_num(inner_sph)
+                inner_sph[inner_sph == 0.0] = 1.0
+
+                # Compute SPH factor residuals
+                inner_res = np.abs((inner_sph - old_inner_sph) / old_inner_sph)
+                inner_res = np.nan_to_num(inner_res)
+
+                # Report maximum SPH factor residual
+                py_printf('NORMAL', 'SPH inner iteration %d:\tres = '
+                                    '%1.3e', j, inner_res.max())
+
+                # Update multi-group cross sections with SPH factors
```

Comment:
Good point, it's a subtle issue. It just updated the comment, let me know if you have any others you'd like fleshed out.